cabal-install/src/Distribution/Client/ProjectConfig/Parsec.hs

```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE OverloadedStrings   #-}
+
+-- | Parsing project configuration.
+
+module Distribution.Client.ProjectConfig.Parsec (
+  -- * Package configuration
+  parseProjectSkeleton,
+
+  -- ** Parsing
+  ParseResult,
+  runParseResult
+  ) where
+
+import Distribution.CabalSpecVersion
+import Distribution.Compat.Prelude
+import Distribution.FieldGrammar
+-- TODO #6101 .Legacy -> ProjectConfigSkeleton should probably be moved here
+import Distribution.Client.ProjectConfig.FieldGrammar (projectConfigFieldGrammar)
+import Distribution.Client.ProjectConfig.Legacy (ProjectConfigSkeleton, ProjectConfigImport)
+import Distribution.Client.ProjectConfig.Types (ProjectConfig)
+import Distribution.Fields.ConfVar                   (parseConditionConfVar)
+import Distribution.Fields.ParseResult
+-- AST type
+import Distribution.Fields (Field, readFields', Name (..))
+import Distribution.Fields.LexerMonad                (LexWarning, toPWarnings)
+import Distribution.PackageDescription.Quirks        (patchQuirks)
+import Distribution.Parsec                           (parsec, simpleParsecBS)
+import Distribution.Parsec.Position                  (Position (..), zeroPos)
+import Distribution.Parsec.Warning                   (PWarnType (..))
+import Distribution.Types.CondTree                   (CondTree (..), CondBranch (..))
+import Distribution.Types.ConfVar                    (ConfVar (..))
+import Distribution.Utils.Generic                    (breakMaybe, fromUTF8BS, toUTF8BS, unfoldrM, validateUTF8)
+
+import qualified Data.ByteString                                   as BS
+import qualified Text.Parsec                                       as P
+
+-- TODO 6101 the following is copied from Distribution.PackageDescription.Parsec.parseGenericPackageDescription
+-- - do we need to use patchQuirks also for cabal.project files? if we do, we can extract a
+-- common preprocessing function (patchQuirks, valid UTF8, ) to avoid code duplication
```

Comment:
I copied this function from `Distribution.PackageDescription.Parsec.parseGenericPackageDescription` because it does some preprocessing of the fields. Do we need to use `patchQuirks` for cabal.project files too? I could extract a common preprocessing function here

Reply:
I don't think that `patchQuirks` is needed for project files. If I understand correctly, it is only for handling a small number of specific packages on Hackage with .cabal files that couldn't be parsed by the Parsec parser when the implementation was changed.

Reply:
Removed it, thanks for the explanation!